### ISSUE TITLE
[bug] plugin doesn't start

### ISSUE THREAD

**Issue by chrisguida** [joelklabo/nostrify#5]
Running on CLN v23.02 installed like this:

https://github.com/Start9Labs/c-lightning-wrapper/blob/22e219d2f4f369fcff8f5a750d63e1c11264400a/Dockerfile#L224-L229

And configured like this:

`plugin=/usr/local/libexec/c-lightning/plugins/nostrify/nostrify.py`

Produces this error whenever the plugin tries to start:

```
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.898Z INFO    plugin-nostrify.py: 
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.898Z INFO    plugin-nostrify.py: Traceback (most recent call last):
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.898Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 668, in _dispatch_notification
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:     self._exec_func(func, request)
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 619, in _exec_func
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:     ret = func(*ba.args, **ba.kwargs)
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:   File \"/usr/local/libexec/c-lightning/plugins/nostrify/nostrify.py\", line 99, in on_warning
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:     send_nostr_event(content, plugin)
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:   File \"/usr/local/libexec/c-lightning/plugins/nostrify/nostrify.py\", line 12, in send_nostr_event
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py:     command = rf'nostril --envelope --sec \"{plugin.secret}\" --content \"{content}\" | websocat {plugin.relay} > /dev/null'
2023-03-07T16:11:29-06:00  2023-03-07T22:11:29.899Z INFO    plugin-nostrify.py: AttributeError: 'Plugin' object has no attribute 'secret'
```

The [README](https://github.com/joelklabo/nostrify#quickstart) says the secret is autogenerated, is this not correct?

**Comment by joelklabo** [comment]
It is, I've been working on getting this to work with v23.02 and just pushed my changes. I think it should be good to go now

**Comment by joelklabo** [comment]
Let me know though if not and I'll look into it.

**Comment by chrisguida** [comment]
Hmm, last time I built from https://github.com/lightningd/plugins, but that doesn't seem to have updated yet.

Building from this repo, I now can't build the plugin at all

```
#0 9.788 Building wheels for collected packages: psutil
#0 9.789   Building wheel for psutil (pyproject.toml): started
#0 9.940   Building wheel for psutil (pyproject.toml): finished with status 'error'
#0 9.944   error: subprocess-exited-with-error
#0 9.944   
#0 9.944   × Building wheel for psutil (pyproject.toml) did not run successfully.
#0 9.944   │ exit code: 1
#0 9.944   ╰─> [45 lines of output]
#0 9.944       running bdist_wheel
#0 9.944       running build
#0 9.944       running build_py
#0 9.944       creating build
#0 9.944       creating build/lib.linux-aarch64-cpython-39
#0 9.944       creating build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_psosx.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_psaix.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_pswindows.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_psposix.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_common.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/__init__.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_compat.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_psbsd.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_pssunos.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       copying psutil/_pslinux.py -> build/lib.linux-aarch64-cpython-39/psutil
#0 9.944       creating build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_testutils.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_process.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_osx.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_unicode.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/__init__.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_bsd.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_connections.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_misc.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/__main__.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_memleaks.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_system.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_windows.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_aix.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/runner.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_sunos.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_contracts.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_posix.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       copying psutil/tests/test_linux.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
#0 9.944       running build_ext
#0 9.944       building 'psutil._psutil_linux' extension
#0 9.944       creating build/temp.linux-aarch64-cpython-39
#0 9.944       creating build/temp.linux-aarch64-cpython-39/psutil
#0 9.944       aarch64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-PN012d/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=594 -DPy_LIMITED_API=0x03060000 -DPSUTIL_LINUX=1 -I/usr/include/python3.9 -c psutil/_psutil_common.c -o build/temp.linux-aarch64-cpython-39/psutil/_psutil_common.o
#0 9.944       psutil/_psutil_common.c:9:10: fatal error: Python.h: No such file or directory
#0 9.944           9 | #include <Python.h>
#0 9.944             |          ^~~~~~~~~~
#0 9.944       compilation terminated.
#0 9.944       error: command '/usr/bin/aarch64-linux-gnu-gcc' failed with exit code 1
#0 9.944       [end of output]
#0 9.944   
#0 9.944   note: This error originates from a subprocess, and is likely not a problem with pip.
#0 9.944   ERROR: Failed building wheel for psutil
#0 9.945 Failed to build psutil
#0 9.945 ERROR: Could not build wheels for psutil, which is required to install pyproject.toml-based projects
------
Dockerfile:236
--------------------
 234 |     # nostrify
 235 |     ADD ./nostrify /usr/local/libexec/c-lightning/plugins/nostrify
 236 | >>> RUN pip install -r /usr/local/libexec/c-lightning/plugins/nostrify/requirements.txt
 237 |     RUN chmod a+x /usr/local/libexec/c-lightning/plugins/nostrify/nostrify.py
 238 |     COPY --from=builder /usr/local/bin/nostril* /usr/local/bin/
--------------------
ERROR: failed to solve: process "/bin/sh -c pip install -r /usr/local/libexec/c-lightning/plugins/nostrify/requirements.txt" did not complete successfully: exit code: 1
```

It seems `psutil` has some kind of undocumented system dependency? I tried installing `build-essential` for the C compiler, because that was missing. Now I get the above error and I'm not sure how to resolve.

**Comment by joelklabo** [comment]
> Hmm, last time I built from https://github.com/lightningd/plugins, but that doesn't seem to have updated yet.
> 
> 
> 
> Building from this repo, I now can't build the plugin at all
> 
> 
> 
> ```
> 
> #0 9.788 Building wheels for collected packages: psutil
> 
> #0 9.789   Building wheel for psutil (pyproject.toml): started
> 
> #0 9.940   Building wheel for psutil (pyproject.toml): finished with status 'error'
> 
> #0 9.944   error: subprocess-exited-with-error
> 
> #0 9.944   
> 
> #0 9.944   × Building wheel for psutil (pyproject.toml) did not run successfully.
> 
> #0 9.944   │ exit code: 1
> 
> #0 9.944   ╰─> [45 lines of output]
> 
> #0 9.944       running bdist_wheel
> 
> #0 9.944       running build
> 
> #0 9.944       running build_py
> 
> #0 9.944       creating build
> 
> #0 9.944       creating build/lib.linux-aarch64-cpython-39
> 
> #0 9.944       creating build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_psosx.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_psaix.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_pswindows.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_psposix.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_common.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/__init__.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_compat.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_psbsd.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_pssunos.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       copying psutil/_pslinux.py -> build/lib.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       creating build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_testutils.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_process.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_osx.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_unicode.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/__init__.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_bsd.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_connections.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_misc.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/__main__.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_memleaks.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_system.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_windows.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_aix.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/runner.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_sunos.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_contracts.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_posix.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       copying psutil/tests/test_linux.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
> 
> #0 9.944       running build_ext
> 
> #0 9.944       building 'psutil._psutil_linux' extension
> 
> #0 9.944       creating build/temp.linux-aarch64-cpython-39
> 
> #0 9.944       creating build/temp.linux-aarch64-cpython-39/psutil
> 
> #0 9.944       aarch64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-PN012d/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=594 -DPy_LIMITED_API=0x03060000 -DPSUTIL_LINUX=1 -I/usr/include/python3.9 -c psutil/_psutil_common.c -o build/temp.linux-aarch64-cpython-39/psutil/_psutil_common.o
> 
> #0 9.944       psutil/_psutil_common.c:9:10: fatal error: Python.h: No such file or directory
> 
> #0 9.944           9 | #include <Python.h>
> 
> #0 9.944             |          ^~~~~~~~~~
> 
> #0 9.944       compilation terminated.
> 
> #0 9.944       error: command '/usr/bin/aarch64-linux-gnu-gcc' failed with exit code 1
> 
> #0 9.944       [end of output]
> 
> #0 9.944   
> 
> #0 9.944   note: This error originates from a subprocess, and is likely not a problem with pip.
> 
> #0 9.944   ERROR: Failed building wheel for psutil
> 
> #0 9.945 Failed to build psutil
> 
> #0 9.945 ERROR: Could not build wheels for psutil, which is required to install pyproject.toml-based projects
> 
> ------
> 
> Dockerfile:236
> 
> --------------------
> 
>  234 |     # nostrify
> 
>  235 |     ADD ./nostrify /usr/local/libexec/c-lightning/plugins/nostrify
> 
>  236 | >>> RUN pip install -r /usr/local/libexec/c-lightning/plugins/nostrify/requirements.txt
> 
>  237 |     RUN chmod a+x /usr/local/libexec/c-lightning/plugins/nostrify/nostrify.py
> 
>  238 |     COPY --from=builder /usr/local/bin/nostril* /usr/local/bin/
> 
> --------------------
> 
> ERROR: failed to solve: process "/bin/sh -c pip install -r /usr/local/libexec/c-lightning/plugins/nostrify/requirements.txt" did not complete successfully: exit code: 1
> 
> ```
> 
> 
> 
> It seems `psutil` has some kind of undocumented system dependency? I tried installing `build-essential` for the C compiler, because that was missing. Now I get the above error and I'm not sure how to resolve.

Are you building on a Mac? I had the same issues actually. Did this happen when you were installing the requirements.txt ?

**Comment by joelklabo** [comment]
@chrisguida only pyln-client is actually required to run the plug-in. Maybe just try pip installing that? The rest are to run the tests

**Comment by chrisguida** [comment]
I'm building in a debian-based dockerfile. looks like the solution was `apt install python3-dev`

**Comment by chrisguida** [comment]
Would be great if this repo had a Dockerfile as an example of what system dependencies should be in place :)

**Comment by chrisguida** [comment]
Or maybe just the dependencies noted in the README

**Comment by joelklabo** [comment]
> Would be great if this repo had a Dockerfile as an example of what system dependencies should be in place :)

It actually does now! I just added that because I kept running into issues building too. And set up Ci. Did you check out master? That's probably the best thing to try

**Comment by chrisguida** [comment]
Excellent, I totally missed that! Thanks!

**Comment by joelklabo** [comment]
> Excellent, I totally missed that! Thanks!

Yeah, I was using codespaces actually. You can do it that way as well

**Comment by chrisguida** [comment]
Ok, now I'm getting this error when the plugin starts:

```
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py: Traceback (most recent call last):
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 668, in _dispatch_notification
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     self._exec_func(func, request)
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 618, in _exec_func
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     ba = self._bind_kwargs(func, params, request)
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 604, in _bind_kwargs
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     ba = sig.bind(**params)
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/lib/python3.9/inspect.py\", line 3062, in bind
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     return self._bind(args, kwargs)
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/lib/python3.9/inspect.py\", line 2977, in _bind
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py:     raise TypeError(msg) from None
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py: TypeError: missing a required argument: 'balance_snapshots'
2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py: 
```

**Comment by joelklabo** [comment]
Hmm, what version of CLN are you using?

**Comment by chrisguida** [comment]
23.02

**Comment by joelklabo** [comment]
Me too I think, trying to cover all these cases with tests

**Comment by joelklabo** [comment]
> Ok, now I'm getting this error when the plugin starts:
> 
> 
> 
> ```
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py: Traceback (most recent call last):
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 668, in _dispatch_notification
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     self._exec_func(func, request)
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 618, in _exec_func
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     ba = self._bind_kwargs(func, params, request)
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 604, in _bind_kwargs
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     ba = sig.bind(**params)
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/lib/python3.9/inspect.py\", line 3062, in bind
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:     return self._bind(args, kwargs)
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.438Z INFO    plugin-nostrify.py:   File \"/usr/lib/python3.9/inspect.py\", line 2977, in _bind
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py:     raise TypeError(msg) from None
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py: TypeError: missing a required argument: 'balance_snapshots'
> 
> 2023-03-10T20:39:48-06:00  2023-03-11T02:39:48.439Z INFO    plugin-nostrify.py: 
> 
> ```

I think I might know what's causing this. I'll let you know when I have a fix. Sorry for the rough start

**Comment by joelklabo** [comment]
Is it possible you're not up to date with master? On this commit specifically 959f8d59efa1e716c01e5b1a5b19b962be3ec7e6

**Comment by joelklabo** [comment]
Hey @chrisguida I've added some tests and removed the `balance_snapshot` event for now altogether so if you pull from master from this point on I hope it will work for you. Please let me know if not though.

**Comment by chrisguida** [comment]
Looks like it started, thanks!

`2023-03-12T14:02:36.767Z INFO    plugin-nostrify.py: Plugin nostrify initialized`

Now to figure out how to test...

**Comment by joelklabo** [comment]
Yeah, I'm working on adding a way to print out the pub key it should be sending events to. But, if you do that make secret thing from the Readme it will give you the private key then you can log in with that band see events hopefully and the pub key so you can follow them

**Comment by joelklabo** [comment]
@chrisguida I just updated the readme with some instructions to follow your plugin here: https://github.com/joelklabo/nostrify/blob/main/README.md

**Comment by chrisguida** [comment]
@joelklabo can you recommend a "tool to convert that to a pubkey you can just follow it with your main Nostr account"? Not sure what encoding is being converted to what here.

**Comment by chrisguida** [comment]
Also, do these events get posted in plaintext on the relay, for anyone who's subscribed to events from the relay to see?

**Comment by joelklabo** [comment]
Yeah, right now. I have a planned update to make it a DM but right now it
just posts like a normal user. But, if you have your own relay you could
post to that and nobody else should see it.

**Comment by joelklabo** [comment]
@chrisguida would you rather have it as a DM? Just curious

**Comment by chrisguida** [comment]
I haven't actually gotten the plugin working yet, so I'm not sure what these events even are / look like. It seems like they could potentially contain sensitive information? If so it seems strange to default the plugin to use a public relay, and also to post the events publicly.

Also, do you have a response for this? https://github.com/joelklabo/nostrify/issues/5#issuecomment-1466649886

**Comment by chrisguida** [comment]
Hmm, now after runnning the `makesecret` command, I can't start nostrify again

```
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.448Z INFO    plugin-nostrify.py: 
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.448Z INFO    plugin-nostrify.py: Traceback (most recent call last):
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 668, in _dispatch_notification
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:     self._exec_func(func, request)
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:   File \"/usr/local/lib/python3.9/dist-packages/pyln/client/plugin.py\", line 619, in _exec_func
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:     ret = func(*ba.args, **ba.kwargs)
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:   File \"/usr/local/libexec/c-lightning/plugins/nostrify/src/nostrify.py\", line 102, in on_warning
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:     send_nostr_event(content)
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:   File \"/usr/local/libexec/c-lightning/plugins/nostrify/src/nostrify.py\", line 10, in send_nostr_event
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py:     command = rf'nostril --envelope --sec \"{plugin.secret}\" --content \"{content}\" | websocat {plugin.relay} > /dev/null'
2023-03-13T13:24:02-06:00  2023-03-13T19:24:02.449Z INFO    plugin-nostrify.py: AttributeError: 'Plugin' object has no attribute 'secret'
```

**Comment by chrisguida** [comment]
Also it appears that the `python3-dev` and `build-essential` packages are doubling the size of my docker image. Probably going to have to leave out `nostrify` for now until I can find a way to include it without expanding the size of the image so radically.

**Comment by joelklabo** [comment]
Yeah, makes sense. Hopefully things will be a little more solid next time you try. Thanks again for bringing up all the issues.

For your question about keys, I use this: https://damus.io/key for convert keys

**Comment by joelklabo** [comment]
> @joelklabo can you recommend a "tool to convert that to a pubkey you can just follow it with your main Nostr account"? Not sure what encoding is being converted to what here.

I think the easiest one is this, https://damus.io/key

**Comment by joelklabo** [comment]
I'm only using docker for repeatable builds and testing BTW, you should be able to just clone the repo and use the nostrify.py executable in your plugins directory, and install requirements.txt. Just FYI, I'm not sure what setup you are working with so maybe that doesn't make sense

**Comment by chrisguida** [comment]
Actually the sizes are:

without nostrify and its dependencies: `238760960`
with `build-essential`: `315423744`
with `python-dev` and `build-essential`: `324515328`
with `nostrify` installed entirely: `487824896`

So, the disk usage appears to be mostly due to whatever dependencies nostrify installs from its `requirements.txt` file. Those appear to be taking up at least 150MB, probably around double that, since the docker images are compressed

**Comment by chrisguida** [comment]
The setup is that we're building a single docker container with lightningd and all associated plugins: https://github.com/Start9Labs/c-lightning-wrapper/blob/feat/new-plugins/Dockerfile

So no, your suggestion would not work.

**Comment by joelklabo** [comment]
Hmm, yeah that doesn't see good

**Comment by chrisguida** [comment]
Would be great if you could help me figure out how to isolate nostrify's build time dependencies from its runtime dependencies, that way we could build nostrify in a previous build stage, then copy only the necessary runtime files over to the final image.

**Comment by chrisguida** [comment]
> > @joelklabo can you recommend a "tool to convert that to a pubkey you can just follow it with your main Nostr account"? Not sure what encoding is being converted to what here.
> 
> I think the easiest one is this, https://damus.io/key

This doesn't appear to understand the format of the secret output by the command `lightning-cli makesecret string=nostr`

**Comment by joelklabo** [comment]
Yeah, sure. I don't think you need any of them actually. Everything in the Dockerfile etc. is basically just so I can run CI tests so you should be able to just clone the GitHub repo into your plugins folder, then in that folder `pip3 install -r requirements.txt` which should just be `pyln-client` so you should already have it actually. If you have any other plugins. You can actually skip that part. 

Basically all you need is the nostrify.py file in your plugins directory and then have nostril, and websocat installed. That should be it

**Comment by joelklabo** [comment]
Yeah, you're right about the key... you can try this: https://snort.social an login with the secret you created, and you will see your npub on your profile. Here's what mine looks like:

<img width="870" alt="image" src="https://user-images.githubusercontent.com/264977/224822066-43e28a1d-4dad-4b8c-8806-661ca36d6148.png">

**Comment by joelklabo** [comment]
I updated the readme to explain that a little better I hope

**Comment by chrisguida** [comment]
> Yeah, sure. I don't think you need any of them actually. Everything in the Dockerfile etc. is basically just so I can run CI tests so you should be able to just clone the GitHub repo into your plugins folder, then in that folder `pip3 install -r requirements.txt` which should just be `pyln-client` so you should already have it actually. If you have any other plugins. You can actually skip that part.
> 
> Basically all you need is the nostrify.py file in your plugins directory and then have nostril, and websocat installed. That should be it

Hmm yes, you are correct. I simply skipped the requirements.txt install and it seems to initialize just fine... Can we update `requirements.txt` to include only `pyln-client` in that case? That seems like the proper solution.

**Comment by chrisguida** [comment]
without `pip install -r requirements.txt`: `240072704`

Only ~1MB larger than without nostrify, that's a major improvement

**Comment by chrisguida** [comment]
I can't seem to connect to your relay from snort.social:

<img width="678" alt="Screenshot 2023-03-13 at 14 29 03" src="https://user-images.githubusercontent.com/7445670/224825028-6125ecc9-aaba-4f7f-806e-bf19921fdeb6.png">

**Comment by chrisguida** [comment]
Actually scratch that, seems to be connected now:
<img width="632" alt="Screenshot 2023-03-13 at 14 30 49" src="https://user-images.githubusercontent.com/7445670/224825294-3132bbef-3913-4219-ae39-2cbe60c728f6.png">

Is there any way to send a test event? This is just a test node, there's no activity on it currently.

**Comment by joelklabo** [comment]
Hmm, I can't try and fix that but did it start working?

**Comment by chrisguida** [comment]
Yes, the relay appears to be working now. Now all I need is to send some events from my nostrify plugin. How can I send some test events?

**Comment by joelklabo** [comment]
With `nostril` on the command line you can create an event pretty easily, like this:

`nostril --envelope --sec "YOUR SECRET" --content "YOUR MESSAGE" | websocat "SOME RELAY (can try wss:// nostr.klabo.blog"`

**Comment by chrisguida** [comment]
Oh actually I hard refreshed snort.social and now I see a bunch of events! Lots of peer connects and disconnects :)

**Comment by chrisguida** [comment]
Seems there was just some friction getting snort.social reading from your relay, not sure what that was about. It all seems to be working now! 🎉

**Comment by chrisguida** [comment]
Ok yeah, seeing all the different types of events coming out of CLN, it definitely seems like these should not be unencrypted on a public relay. Is there an ETA on DMs?

**Comment by chrisguida** [comment]
I could also just default this to the user's nostr relay and warn them if they try to use a public one

**Comment by joelklabo** [comment]
Yeah, no ETA yet but I can pick that up next if you think you'd use it. Maybe a few days though

**Comment by joelklabo** [comment]
Yeah, another thing I need to add is silencing ones you don't want to see, it's a lot of information

**Comment by joelklabo** [comment]
Feel free to make any feature requests you have as issues and I'll try to get to them if I can. Thanks for working through that!

**Comment by chrisguida** [comment]
Yeah, I think DMs would be a necessary feature. As a stopgap I'll implement the behavior I mentioned above

**Comment by joelklabo** [comment]
> Yeah, I think DMs would be a necessary feature. As a stopgap I'll implement the behavior I mentioned above

Cool, working that now

**Comment by joelklabo** [comment]
FYI @chrisguida I just updated to be able to send DMs. It also removes the websocat and nostril dependencies. Let me know if the README doesn't make sense but I tried to cover it there
